--- ppp-2.4.2/pppd/Makefile.linux.no_strip	2004-09-16 12:00:08.550956895 +0200
+++ ppp-2.4.2/pppd/Makefile.linux	2004-09-16 12:00:08.651942974 +0200
@@ -99,7 +99,7 @@
 CFLAGS	+= -DUSE_SRP -DOPENSSL -I/usr/local/ssl/include
 LIBS	+= -lsrp -L/usr/local/ssl/lib -lcrypto
 TARGETS	+= srp-entry
-EXTRAINSTALL = $(INSTALL) -s -c -m 555 srp-entry $(BINDIR)/srp-entry
+EXTRAINSTALL = $(INSTALL) -c -m 555 srp-entry $(BINDIR)/srp-entry
 MANPAGES += srp-entry.8
 EXTRACLEAN += srp-entry.o
 NEEDDES=y
@@ -200,7 +200,7 @@
 install: pppd
 	mkdir -p $(BINDIR) $(MANDIR)
 	$(EXTRAINSTALL)
-	$(INSTALL) -s -c -m 555 pppd $(BINDIR)/pppd
+	$(INSTALL) -c -m 555 pppd $(BINDIR)/pppd
 	if chgrp pppusers $(BINDIR)/pppd 2>/dev/null; then \
 	  chmod o-rx,u+s $(BINDIR)/pppd; fi
 	$(INSTALL) -c -m 444 pppd.8 $(MANDIR)/man8
--- ppp-2.4.2/pppd/plugins/radius/Makefile.linux.no_strip	2004-09-16 12:00:08.518961306 +0200
+++ ppp-2.4.2/pppd/plugins/radius/Makefile.linux	2004-09-16 12:00:08.643944077 +0200
@@ -30,9 +30,9 @@
 install: all
 	$(MAKE) $(MFLAGS) -C radiusclient install
 	$(INSTALL) -d -m 755 $(LIBDIR)
-	$(INSTALL) -s -c -m 755 radius.so $(LIBDIR)
-	$(INSTALL) -s -c -m 755 radattr.so $(LIBDIR)
-	$(INSTALL) -s -c -m 755 radrealms.so $(LIBDIR)
+	$(INSTALL) -c -m 755 radius.so $(LIBDIR)
+	$(INSTALL) -c -m 755 radattr.so $(LIBDIR)
+	$(INSTALL) -c -m 755 radrealms.so $(LIBDIR)
 	$(INSTALL) -c -m 444 pppd-radius.8 $(MANDIR)/man8
 	$(INSTALL) -c -m 444 pppd-radattr.8 $(MANDIR)/man8
 
--- ppp-2.4.2/pppd/plugins/rp-pppoe/Makefile.linux.no_strip	2004-09-16 12:00:08.518961306 +0200
+++ ppp-2.4.2/pppd/plugins/rp-pppoe/Makefile.linux	2004-09-16 12:00:08.650943112 +0200
@@ -25,7 +25,7 @@
 
 install: all
 	$(INSTALL) -d -m 755 $(LIBDIR)
-	$(INSTALL) -s -c -m 4550 rp-pppoe.so $(LIBDIR)
+	$(INSTALL) -c -m 4550 rp-pppoe.so $(LIBDIR)
 
 clean:
 	rm -f *.o *.so
--- ppp-2.4.2/pppd/Makefile.linux.lib64.no_strip	2004-09-16 12:00:08.516961582 +0200
+++ ppp-2.4.2/pppd/Makefile.linux.lib64	2004-09-16 12:00:08.641944353 +0200
@@ -98,7 +98,7 @@
 CFLAGS	+= -DUSE_SRP -DOPENSSL -I/usr/local/ssl/include
 LIBS	+= -lsrp -L/usr/local/ssl/lib -lcrypto
 TARGETS	+= srp-entry
-EXTRAINSTALL = $(INSTALL) -s -c -m 555 srp-entry $(BINDIR)/srp-entry
+EXTRAINSTALL = $(INSTALL) -c -m 555 srp-entry $(BINDIR)/srp-entry
 MANPAGES += srp-entry.8
 EXTRACLEAN += srp-entry.o
 NEEDDES=y
@@ -199,7 +199,7 @@
 install: pppd
 	mkdir -p $(BINDIR) $(MANDIR)
 	$(EXTRAINSTALL)
-	$(INSTALL) -s -c -m 555 pppd $(BINDIR)/pppd
+	$(INSTALL) -c -m 555 pppd $(BINDIR)/pppd
 	if chgrp pppusers $(BINDIR)/pppd 2>/dev/null; then \
 	  chmod o-rx,u+s $(BINDIR)/pppd; fi
 	$(INSTALL) -c -m 444 pppd.8 $(MANDIR)/man8
--- ppp-2.4.2/pppd/Makefile.linux.pie.no_strip	2004-09-16 12:00:08.536958825 +0200
+++ ppp-2.4.2/pppd/Makefile.linux.pie	2004-09-16 12:00:08.651942974 +0200
@@ -99,7 +99,7 @@
 CFLAGS	+= -DUSE_SRP -DOPENSSL -I/usr/local/ssl/include
 LIBS	+= -lsrp -L/usr/local/ssl/lib -lcrypto
 TARGETS	+= srp-entry
-EXTRAINSTALL = $(INSTALL) -s -c -m 555 srp-entry $(BINDIR)/srp-entry
+EXTRAINSTALL = $(INSTALL) -c -m 555 srp-entry $(BINDIR)/srp-entry
 MANPAGES += srp-entry.8
 EXTRACLEAN += srp-entry.o
 NEEDDES=y
@@ -200,7 +200,7 @@
 install: pppd
 	mkdir -p $(BINDIR) $(MANDIR)
 	$(EXTRAINSTALL)
-	$(INSTALL) -s -c -m 555 pppd $(BINDIR)/pppd
+	$(INSTALL) -c -m 555 pppd $(BINDIR)/pppd
 	if chgrp pppusers $(BINDIR)/pppd 2>/dev/null; then \
 	  chmod o-rx,u+s $(BINDIR)/pppd; fi
 	$(INSTALL) -c -m 444 pppd.8 $(MANDIR)/man8
--- ppp-2.4.2/pppdump/Makefile.linux.no_strip	2004-09-16 12:00:08.516961582 +0200
+++ ppp-2.4.2/pppdump/Makefile.linux	2004-09-16 12:00:08.653942699 +0200
@@ -13,5 +13,5 @@
 
 install:
 	mkdir -p $(BINDIR) $(MANDIR)/man8
-	$(INSTALL) -s -c pppdump $(BINDIR)
+	$(INSTALL) -c pppdump $(BINDIR)
 	$(INSTALL) -c -m 444 pppdump.8 $(MANDIR)/man8
--- ppp-2.4.2/pppstats/Makefile.linux.no_strip	2004-09-16 12:00:08.517961444 +0200
+++ ppp-2.4.2/pppstats/Makefile.linux	2004-09-16 12:00:08.654942561 +0200
@@ -19,7 +19,7 @@
 
 install: pppstats
 	-mkdir -p $(MANDIR)/man8
-	$(INSTALL) -s -c pppstats $(BINDIR)/pppstats
+	$(INSTALL) -c pppstats $(BINDIR)/pppstats
 	$(INSTALL) -c -m 444 pppstats.8 $(MANDIR)/man8/pppstats.8
 
 pppstats: $(PPPSTATSRCS)
--- ppp-2.4.2/chat/Makefile.linux.make.no_strip	2004-01-13 04:57:55.000000000 +0100
+++ ppp-2.4.2/chat/Makefile.linux.make	2004-09-16 12:00:08.655942423 +0200
@@ -21,7 +21,7 @@
 
 install: chat
 	mkdir -p $(BINDIR)
-	$(INSTALL) -s -c chat $(BINDIR)
+	$(INSTALL) -c chat $(BINDIR)
 	$(INSTALL) -c -m 644 chat.8 $(MANDIR)/man8
 
 clean:
--- ppp-2.4.2/chat/Makefile.linux.no_strip	2004-09-16 12:00:08.551956757 +0200
+++ ppp-2.4.2/chat/Makefile.linux	2004-09-16 12:00:08.655942423 +0200
@@ -21,7 +21,7 @@
 
 install: chat
 	mkdir -p $(BINDIR)
-	$(INSTALL) -s -c chat $(BINDIR)
+	$(INSTALL) -c chat $(BINDIR)
 	$(INSTALL) -c -m 644 chat.8 $(MANDIR)/man8
 
 clean:
--- ppp-2.4.2/chat/Makefile.linux.pie.no_strip	2004-09-16 12:00:08.515961719 +0200
+++ ppp-2.4.2/chat/Makefile.linux.pie	2004-09-16 12:00:08.655942423 +0200
@@ -21,7 +21,7 @@
 
 install: chat
 	mkdir -p $(BINDIR)
-	$(INSTALL) -s -c chat $(BINDIR)
+	$(INSTALL) -c chat $(BINDIR)
 	$(INSTALL) -c -m 644 chat.8 $(MANDIR)/man8
 
 clean:
