--- ppp/scripts/ppp-on-rsh.old	2005-07-19 09:16:55.000000000 -0500
+++ ppp/scripts/ppp-on-rsh	2005-07-19 09:16:39.000000000 -0500
@@ -26,7 +26,7 @@
 # For this example, we assume that pppd on both local and remote
 # machines reside in the same place, /usr/local/bin/pppd
 #
-PPPD_LOC=/usr/local/bin/pppd
+PPPD_LOC=/usr/sbin/pppd
 
 #
 # The location of local options file (where rsh client is running).
--- ppp/scripts/secure-card.old	2004-02-01 21:36:46.000000000 -0600
+++ ppp/scripts/secure-card	2005-07-19 09:14:19.000000000 -0500
@@ -1,4 +1,4 @@
-#!/usr/local/bin/expect -f
+#!/usr/bin/expect -f
 #
 # This  script was  written  by  Jim Isaacson  <jcisaac@crl.com>.  It is
 # designed to work  as a script to use the  SecureCARD(tm) device.  This
--- ppp/scripts/ppp-on-ssh.old	2000-04-15 04:49:42.000000000 -0500
+++ ppp/scripts/ppp-on-ssh	2005-07-19 09:17:25.000000000 -0500
@@ -26,7 +26,7 @@
 # For this example, we assume that pppd on both local and remote
 # machines reside in the same place, /usr/local/bin/pppd
 #
-PPPD_LOC=/usr/local/bin/pppd
+PPPD_LOC=/usr/sbin/pppd
 
 #
 # The location of local options file (where ssh client is running).
@@ -52,7 +52,7 @@
 #
 # The location of ssh client on the local machine
 #
-SSH_LOC=/usr/local/bin/ssh
+SSH_LOC=/usr/bin/ssh
 
 export PPPD_LOC PPPD_LOC_OPT PPPD_REM_OPT PPPD_RHOST SSH_LOC
 
--- ppp/configure.old	2004-11-06 05:36:54.000000000 -0600
+++ ppp/configure	2005-07-19 09:24:29.000000000 -0500
@@ -2,7 +2,7 @@
 # $Id: configure,v 1.35 2004/11/06 11:36:54 paulus Exp $
 
 # Where to install stuff by default
-DESTDIR=/usr/local
+DESTDIR=/usr
 SYSCONF=/etc
 
 #  if [ -d /NextApps ]; then
--- ppp-2.4.3/pppd/Makefile.linux.old	2005-07-19 09:22:21.000000000 -0500
+++ ppp-2.4.3/pppd/Makefile.linux	2005-07-19 09:30:58.000000000 -0500
@@ -96,8 +96,8 @@
 
 # EAP SRP-SHA1
 ifdef USE_SRP
-CFLAGS	+= -DUSE_SRP -DOPENSSL -I/usr/local/ssl/include
-LIBS	+= -lsrp -L/usr/local/ssl/lib -lcrypto
+CFLAGS	+= -DUSE_SRP -DOPENSSL -I/usr/include/openssl
+LIBS	+= -lsrp -L/usr/lib -lcrypto
 TARGETS	+= srp-entry
 EXTRAINSTALL = $(INSTALL) -c -m 555 srp-entry $(BINDIR)/srp-entry
 MANPAGES += srp-entry.8
