I think this is the right fix, but ICBW.
--- ppp-2.4.2/pppd/demand.c	2004-03-10 19:12:16.000000000 -0500
+++ ppp-2.4.2/pppd/demand.c	2004-03-10 19:12:12.000000000 -0500
@@ -50,7 +50,7 @@
 #include <sys/socket.h>
 #ifdef PPP_FILTER
 #include <net/if.h>
-#include <net/bpf.h>
+#include <pcap-bpf.h>
 #include <pcap.h>
 #endif
 
--- ppp-2.4.2/pppd/sys-linux.c	2004-03-10 19:11:56.000000000 -0500
+++ ppp-2.4.2/pppd/sys-linux.c	2004-03-10 19:12:05.000000000 -0500
@@ -141,7 +141,7 @@
 #endif /* IPX_CHANGE */
 
 #ifdef PPP_FILTER
-#include <net/bpf.h>
+#include <pcap-bpf.h>
 #include <linux/filter.h>
 #endif /* PPP_FILTER */
 
